java-checks/src/main/java/org/sonar/java/checks/DateTimeFormatterMismatchCheck.java

```diff
@@ -0,0 +1,221 @@
+      invocation.accept(visitor);
+      if (!visitor.usesWeek || !visitor.usesYear) {
+        return;
+      }
+      if (visitor.usesWeekBasedYear && !visitor.usesWeekOfWeekBasedYear) {
+        reportIssue(visitor.primary, CHANGE_YEAR_FORMAT_TO_CHRONOFIELD_MESSAGE, visitor.locations, null);
+      } else if (!visitor.usesWeekBasedYear && visitor.usesWeekOfWeekBasedYear) {
+        reportIssue(visitor.primary, CHANGE_YEAR_FORMAT_WEEK_BASED_MESSAGE, visitor.locations, null);
+      }
+    }
+  }
+
+  /**
+   * We currently only check formatters initialized with a String literal or the value of a final variable.
+   */
+  private void visitPattern(MethodInvocationTree invocation) {
+    Arguments arguments = invocation.arguments();
+    ExpressionTree argument = arguments.get(0);
+    if (argument.is(Tree.Kind.STRING_LITERAL)) {
+      String pattern = ((LiteralTree) argument).value();
+      if (isInfringingPattern(pattern)) {
+        reportIssue(invocation, CHANGE_YEAR_FORMAT_WEEK_BASED_MESSAGE);
+      }
+    } else if (argument.is(Tree.Kind.IDENTIFIER)) {
+      IdentifierTree identifier = (IdentifierTree) argument;
+      Symbol symbol = identifier.symbol();
+      if (!symbol.isFinal()) {
+        return;
+      }
+      Tree declaration = symbol.declaration();
+      Optional<String> pattern = getStringLiteralValue(declaration);
+      if (pattern.isPresent() && isInfringingPattern(pattern.get())) {
+        reportIssue(invocation, CHANGE_YEAR_FORMAT_WEEK_BASED_MESSAGE);
+      }
+    }
+  }
+
+  private static Optional<String> getStringLiteralValue(Tree declaration) {
+    if (declaration == null || !declaration.is(Tree.Kind.VARIABLE)) {
+      return Optional.empty();
+    }
+    VariableTree variable = (VariableTree) declaration;
+    ExpressionTree initializer = variable.initializer();
+    if (initializer == null || !initializer.is(Tree.Kind.STRING_LITERAL)) {
+      return Optional.empty();
+    }
+    return Optional.of(((LiteralTree) initializer).value());
+  }
+
+  private static boolean isInfringingPattern(String pattern) {
+    return WEEK_PATTERN.matcher(pattern).matches() && YEAR_OF_ERA_PATTERN.matcher(pattern).matches();
+  }
+
+  private static boolean refersToWeek(ExpressionTree argument) {
+    return isChronoFieldWeek(argument) || isWeekOfWeekBasedYearUsed(argument);
+  }
+
+  private static boolean isWeekOfWeekBasedYearUsed(ExpressionTree argument) {
+    if (argument.is(Tree.Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree call = (MethodInvocationTree) argument;
+      return WEEK_OF_WEEK_BASED_YEAR_MATCHER.matches(call);
+    }
+    return false;
+  }
+
+  private static boolean isChronoFieldWeek(ExpressionTree argument) {
+    if (argument.is(Tree.Kind.MEMBER_SELECT)) {
+      MemberSelectExpressionTree select = (MemberSelectExpressionTree) argument;
+      IdentifierTree identifier = select.identifier();
+      return identifier.name().equals("ALIGNED_WEEK_OF_YEAR");
```

Comment:
For safety, we could also check the experssion underlying type:
`return select.symbolType().is("java.time.temporal.ChronoField") && identifier.name().equals("ALIGNED_WEEK_OF_YEAR");`